### PR TITLE
feat(forge): fuzz & invariant config

### PR DESCRIPTION
**Pull request by rkrasiuk**
create fuzz & invariant testing configuration discussed in #1572

### CODE REVIEW

File: config/README.md [hunk 1]
```diff
@@ -160,8 +152,30 @@ revert_strings = "default"
...
+fail_on_revert = false
+call_override = false
+dictionary_weight = 80
+include_storage = true
+include_push_bytes = true
+
+[fmt]
+line_length = 100
+tab_width = 2
+bracket_spacing = true
```
Comment (mattsse) [inline config/README.md:178]: awesome

File: config/src/lib.rs [hunk 2]
```diff
@@ -362,7 +349,12 @@ impl Config {
     pub const PROFILE_SECTION: &'static str = "profile";
 
     /// Standalone sections in the config which get integrated into the selected profile
-    pub const STANDALONE_SECTIONS: &'static [&'static str] = &["rpc_endpoints", "etherscan", "fmt"];
+    pub const STANDALONE_SECTIONS: &'static [&'static str] =
+        &["rpc_endpoints", "etherscan", "fmt", "fuzz", "invariant"];
+
+    /// Mapping of fallback standalone sections. See [`FallbackProfileProvider`]
+    pub const STANDALONE_FALLBACK_SECTIONS: Lazy<HashMap<&'static str, &'static str>> =
```
Comment (mattsse) [inline config/src/lib.rs:356]: clippy: make static

File: config/src/lib.rs [hunk 3]
```diff
@@ -2299,6 +2312,44 @@ impl<P: Provider> Provider for OptionalStrictProfileProvider<P> {
     }
 }
 
+/// Extracts the profile from the `profile` key and sets unset values according to the fallback
+/// provider
+struct FallbackProfileProvider<P> {
+    provider: P,
+    profile: Profile,
+    fallback: Profile,
+}
```
Comment (mattsse) [inline config/src/lib.rs:2321]: let's move this to providers.rs

**Review by mattsse — CHANGES_REQUESTED** [review]
nice, this looks very good!

File: evm/src/fuzz/invariant/executor.rs [hunk 4]
```diff
@@ -147,7 +150,14 @@ impl<'a> InvariantExecutor<'a> {
...
                         call_result.state_changeset.to_owned().expect("to have a state changeset.");
 
-                    collect_data(&mut state_changeset, sender, &call_result, fuzz_state.clone());
+                    collect_data(
+                        &mut state_changeset,
+                        sender,
+                        &call_result,
+                        fuzz_state.clone(),
+                        self.config.include_push_bytes,
+                        self.config.include_push_bytes,
```
Comment (joshieDo) [inline evm/src/fuzz/invariant/executor.rs:159]: `include_storage` ?

**Comment by joshieDo** [comment]
I feel like maybe we need to extend `dictionary_weight`.

We do use weights for more stuff (wherever there is an `Union` basically). Either here, on a quick follow-up

File: evm/src/fuzz/invariant/executor.rs [hunk 5]
```diff
@@ -230,23 +239,28 @@ impl<'a> InvariantExecutor<'a> {
         }
 
         // Stores fuzz state for use with [fuzz_calldata_from_state].
-        let fuzz_state: EvmFuzzState = build_initial_state(self.executor.backend().mem_db());
+        let fuzz_state: EvmFuzzState =
+            build_initial_state(self.executor.backend().mem_db(), self.config.include_storage);
```
Comment (joshieDo) [inline evm/src/fuzz/invariant/executor.rs:243]: just leaving a note, that if this gets merged https://github.com/foundry-rs/foundry/pull/2929 , we should update `build_initial_state` to reflect that

File: evm/src/fuzz/invariant/executor.rs (hunk 4 above)
Comment (rkrasiuk) [reply evm/src/fuzz/invariant/executor.rs:159]: good catch

**Comment by rkrasiuk** [comment]
@joshieDo hmm, from UX standpoint, do u think we should just keep adding `_weight` suffixed entries?

**Comment by rkrasiuk** [comment]
@mattsse since this is a breaking change, do you think it's smth we should fix w/ `forge config --fix`?

**Comment by mattsse** [comment]
given that invariant and fmt where just added and not part of a release yet, this should be fine imo

**Pull request merged**